### PR TITLE
Add Path@Penn-based Course Schedule page

### PR DESCRIPTION
**Pull request by anli5005**
This PR:
* Adds `PathAtPennNetworkManager` for fetching courses from good ol' [Path@Penn](https://path.at.upenn.edu/?autologin=deepLink=/student/landing)
* Replaces the Course Schedule page with a new SwiftUI-based one
* Removes a whole bunch of code related to the legacy PennInTouch courses implementation

Stuff that still needs/may need to be implemented:
* Uploading courses for recommendation purposes and the relevant privacy setting for that
* Detail/map view when a course is tapped
* Gracefully handling Path@Penn login failures with `PennLoginController` or something to that effect
  * For now the app has an error message - if you get this, log out/log back in

Some screenshots (with appropriately redacted info, given that publishing my exact whereabouts for the world to see might not be the best idea):

IMG_0291
IMG_0290

Obligatory disclaimer that this is my first contribution and I may or may not be doing things very wrong.

### CODE REVIEW

File: PennMobile/General/Generics/GenericControllers.swift [hunk 1]
```diff
@@ -106,3 +107,43 @@ class GenericViewController: UIViewController, Trackable {
...
     }
 }
+
+/// ``GenericViewController`` instance that uses a SwiftUI view for its content.
+///
+/// To use this view controller, subclass and override ``content`` and ``tabTitle``.
+class GenericSwiftUIViewController<Content: View>: GenericViewController {
+    /// Content of the view controller.
+    open var content: Content {
+        fatalError("makeSwiftUIView() not implemented")
```
Comment (01jongmin) [inline PennMobile/General/Generics/GenericControllers.swift:117]: Nice 👍  We should use this for our dining view controller and future wrapped version of UIKit Controllers

File: PennMobile/Login/PathAtPennNetworkManager.swift [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    static private let oauthURL = URL(string: "https://idp.pennkey.upenn.edu/idp/profile/oidc/authorize?response_type=code&scope=openid+email+profile&client_id=courses.upenn.edu%2Fsam_rMReby8Vl3M1BiyVWMAT&redirect_uri=https%3A%2F%2Fcourses.upenn.edu%2Fsam%2Fcodetotoken")!
+
+    /// Fetches and returns a Path@Penn auth token.
+    func getToken() async throws -> String {
+        let (data, _) = try await URLSession.shared.data(from: PathAtPennNetworkManager.oauthURL)
+        let str = try String(data: data, encoding: .utf8).unwrap(orThrow: PathAtPennError.corruptString)
+        let matches = str.getMatches(for: "value: \"(.*)\"")
+        if matches.isEmpty {
+            print(str)
+        }
```
Comment (01jongmin) [inline PennMobile/Login/PathAtPennNetworkManager.swift:45]: maybe remove this print statement

File: PennMobile/Courses/Models/CoursesModel.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+//  PennMobile
+//
+//  Created by Anthony Li on 9/30/22.
+//  Copyright © 2022 PennLabs. All rights reserved.
+//
+
+import Foundation
+
+/// Manages the fetching of courses.
+class CoursesModel: ObservableObject {
```
Comment (01jongmin) [inline PennMobile/Courses/Models/CoursesModel.swift:12]: Our code base usually names ObservableObjects as __ViewModel, so to fit the style, maybe renaming it to CoursesViewModel might be ideal

File: PennMobile/Courses/Views/CoursesView.swift [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                WeekView(courses: courses.filter {
+                    if let start = $0.startDate, let end = $0.endDate {
+                        return start <= date && date <= end.addingTimeInterval(24 * 60 * 60)
+                    } else {
+                        return false
+                    }
+                })
+            }
+        }.task {
+            _ = try? await coursesModel.fetchCourses()
```
Comment (01jongmin) [inline PennMobile/Courses/Views/CoursesView.swift:72]: Does the linter throw a warning if you try to just use `try? await coursesModel.fetchCourses()` without the `_ =`?

File: PennMobile/Courses/Models/Course.swift [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                        return MeetingTime(weekday: weekday, startTime: start, endTime: end)
+                    } catch let error {
+                        print("Couldn't parse start/end times: \(error)")
+                        return nil
+                    }
+                }
+            } catch let error {
+                meetingTimes = []
+                print("Couldn't parse meeting times: \(error)")
+            }
```
Comment (01jongmin) [inline PennMobile/Courses/Models/Course.swift:143]: is the nested `do` `catch` block necessary here as well?

File: PennMobile/Login/PathAtPennNetworkManager.swift [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            let (srcdb, descriptors) = $0
+
+            let crns = descriptors.compactMap {
+                $0.split(separator: "|").first
+            }
+
+            return try await crns.asyncMap { crn in
+                try await self.fetchCourse(srcdb: srcdb, crn: String(crn))
+            }.compactMap { $0 }
+        }.flatMap { $0 }
```
Comment (01jongmin) [inline PennMobile/Login/PathAtPennNetworkManager.swift:157]: Nice async function chaining 👍

**Review by 01jongmin — CHANGES_REQUESTED** [review]
Great job submitting your first PR to this repo! Just a couple of changes!

File: PennMobile/General/Generics/GenericControllers.swift (hunk 1 above)
Comment (anli5005) [reply PennMobile/General/Generics/GenericControllers.swift:117]: Yeah, was considering rewriting DiningViewController as well, but we could probably do that in a separate PR.

File: PennMobile/Login/PathAtPennNetworkManager.swift (hunk 2 above)
Comment (anli5005) [reply PennMobile/Login/PathAtPennNetworkManager.swift:45]: Oh yeah, I should remove that. That was originally there for debugging cases where Path@Penn wouldn't return a token, but we'll need to resolve that later with PennLoginController, probably

File: PennMobile/Courses/Views/CoursesView.swift (hunk 4 above)
Comment (anli5005) [reply PennMobile/Courses/Views/CoursesView.swift:72]: Yeah, it does. I'll just slap a `@discardableResult` on it though.

Comment (anli5005) [reply PennMobile/Courses/Views/CoursesView.swift:72]: Never mind, it looks like that doesn't silence the linter. The intent is for the return value to be used for uploading courses later, so hopefully we can leave this as is.

File: PennMobile/Login/PathAtPennNetworkManager.swift (hunk 6 above)
Comment (anli5005) [reply PennMobile/Login/PathAtPennNetworkManager.swift:157]: Thanks!

File: PennMobile/Courses/Models/Course.swift (hunk 5 above)
Comment (anli5005) [reply PennMobile/Courses/Models/Course.swift:143]: Yeah, it's probably cleaner without it.

File: PennMobile/General/Generics/GenericControllers.swift (hunk 1 above)
Comment (01jongmin) [reply PennMobile/General/Generics/GenericControllers.swift:117]: Yup, we could complete it on another PR.

**Review by 01jongmin — APPROVED** [review]
LGTM! Feel free to merge it in!

**Pull request merged**